network/tcp.go

```diff
@@ -0,0 +1,411 @@
+package network
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dedis/cothority/log"
+)
+
+// NewTCPRouter returns a new Router using TCPHost as the underlying Host.
+func NewTCPRouter(sid *ServerIdentity) (*Router, error) {
+	h, err := NewTCPHost(sid.Address)
+	if err != nil {
+		return nil, err
+	}
+	r := NewRouter(sid, h)
+	return r, nil
+}
+
+// TCPConn implements the Conn interface using plain, unencrypted TCP.
+type TCPConn struct {
+	// The name of the endpoint we are connected to.
+	endpoint Address
+
+	// The connection used
+	conn net.Conn
+
+	// closed indicator
+	closed    bool
+	closedMut sync.Mutex
+	// So we only handle one receiving packet at a time
+	receiveMutex sync.Mutex
+	// So we only handle one sending packet at a time
+	sendMutex sync.Mutex
+
+	counterSafe
+}
+
+// NewTCPConn will open a TCPConn to the given address.
+// In case of an error it returns a nil TCPConn and the error.
+func NewTCPConn(addr Address) (*TCPConn, error) {
+	var err error
+	var conn net.Conn
+	netAddr := addr.NetworkAddress()
+	for i := 0; i < MaxRetryConnect; i++ {
+		conn, err = net.Dial("tcp", netAddr)
+		if err != nil {
+			time.Sleep(WaitRetry)
+		} else {
+			break
+		}
+		time.Sleep(WaitRetry)
+	}
+	if conn == nil {
+		return nil, fmt.Errorf("Could not connect to %s: %s", addr, err)
+	}
+	c := TCPConn{
+		endpoint: addr,
+		conn:     conn,
+	}
+	return &c, err
+}
+
+// Receive calls the receive routine to get the bytes from the connection then
+// it decodes the buffer.
```

Comment:
Change to: 
```
// Receive receives the raw bytes from the connection and
// tries decoding them into a `Packet`.
```
(To my opinion a caller from outside shouldn't care about the "receive routine")